### PR TITLE
Add `min_payment_amt` configuration option to allow setting minimum payment amount (#612)

### PR DESCRIPTION
**Pull request by Vlad1mir-D**
Feature: Add `min_payment_amt` configuration option to allow setting minimum payment amount (#612)

---
name: Pull Request
about: Create a pull request to make a contribution
labels: [enhancement](https://github.com/tezos-reward-distributor-organization/tezos-reward-distributor/labels/enhancement)

---
IMPORTANT NOTICE:
I read and understood the [guidelines for contributions to the TRD](https://tezos-reward-distributor-organization.github.io/tezos-reward-distributor/contributors.html). The contribution may qualify for being compensated by the TRD grant if approved by the maintainers.

This PR resolves the issue 612. The following steps were performed:

* **Solution**: Added support for `min_payment_amt` configuration option similar to [`payment_requirements.minimum_amount`](https://github.com/kalouo/breadcrumbs/wiki/2.-Configuration#payment_requirements) option which is implemented in the [Breadcrumbs](https://github.com/kalouo/breadcrumbs).

* **Implementation**: At first I did [attempt](https://github.com/Vlad1mir-D/tezos-reward-distributor/commit/1c83073573391cbbaf9af909bff2f8587a5111b3) to implement this option in [`calc/calculate_phase_final.py`](https://github.com/Vlad1mir-D/tezos-reward-distributor/commit/1c83073573391cbbaf9af909bff2f8587a5111b3#diff-5a212ccad57abfed4cbe7db9c1df13af42eff52ee224a3728e978d3f8ff53afd) but went nowhere. Then I realized it will be much easier to make a recursive call for [`calculate()`](https://github.com/tezos-reward-distributor-organization/tezos-reward-distributor/commit/0b667e179447e31f47eb72edf41773057c7d1a08#diff-4bd82803d5e7462d2f2f3d9d1e120aa6e6e6ef988bfc556e91ac7b15b1ec2820R155) in [`calc/phased_payment_calculator.py`](https://github.com/tezos-reward-distributor-organization/tezos-reward-distributor/commit/0b667e179447e31f47eb72edf41773057c7d1a08#diff-4bd82803d5e7462d2f2f3d9d1e120aa6e6e6ef988bfc556e91ac7b15b1ec2820) and this seems to be the right choice in my opinion.

* **Performed tests**:
  * Executed simulation runs for cycles 1-518 and compared results with calculations made without this feature being enabled
  * Using in production starting from cycle 500

* **Check list**:

  - [ ] I extended the Github Actions CI test units with the corresponding tests for this new feature (if needed).
  - [X] I extended the Sphinx documentation (if needed).
  - [ ] I extended the help section (if needed).
  - [ ] I changed the README file (if needed).
  - [ ] I created a new issue if there is further work left to be done (if needed).

**Work effort**: 12h

### CODE REVIEW

**Comment by jdsika** [comment]
It is planned for next week! I am still on vacation:))

**Comment by Vlad1mir-D** [comment]
> It is planned for next week! I am still on vacation:))

Oh, didn't mean to bother you. Thanks for this reply!

File: src/configure.py [hunk 1]
```diff
@@ -202,6 +204,19 @@ def onmindelegation(input):
     fsm.go()
 
 
+def onminpayment(input):
+    try:
+        if not input:
+            input = "0"
+        global parser
+        parser.set(MIN_PAYMENT_AMT, float(input))
+        parser.validate_service_fee(parser.get_conf_obj())
```
Comment (jdsika) [inline src/configure.py:213]: Why validate service fee here? And is the text for the exception correct?

File: tests/smoke/test_launch_states.py [hunk 2]
```diff
@@ -142,7 +143,7 @@ def test_wrong_api_base_url(args):
     MagicMock(return_value=""),
 )
 def test_wrong_node_end_point(args):
-    args.node_endpoint = "https://testnet-tezos.giganode.io:4433"
```
Comment (jdsika) [inline tests/smoke/test_launch_states.py:145]: I think the 4433 is on purpose as it should test a "wrong endpoint"?

File: src/configure.py (hunk 1 above)
Comment (Vlad1mir-D) [reply src/configure.py]: This is dumb copy-paste from `onmindelegation` which seems to be the same dumb copy-paste from `onservicefee`.
Fixed both `onmindelegation` and `onminpayment`.

**Comment by Vlad1mir-D** [comment]
@jdsika Successfully resolved all notices, thank you for your time!

**Comment by Vlad1mir-D** [comment]
@jdsika What do you think? Do I need to fix/improve any other parts?

**Comment by Vlad1mir-D** [comment]
@jdsika I understand that TRD is an open-source software and there are no obligations exist for maintainers but I have a bunch of features developed for TRD and would like to donate them properly to the TRD community.
I'm not attempting to push you in any way but I want to know if there are some shortcuts I can use to speed-up the MR review process.
Thank you for your time!

File: docs/configuration.rst [hunk 3]
```diff
@@ -126,6 +126,13 @@ Available configuration parameters are:
 
     min_delegation_amt : 10
   
+**min_payment_amt**
+  A minimum payment amount can be set here. If this value is set to 10, 10 tez are required as minimum. Inherits behavior of excluded delegates set for *min_delegation_amt*.
```
Comment (jdsika) [inline docs/configuration.rst:130]: I am having trouble understanding the "inherits behavior [...]" part. What exactly is inherited here? I am setting a value elsewhere and it is copied to min_payment_amt?

Btw I think the abbreviations like "amt" for "amount" are a curse in the code. In the config we are basically forced to use it due to consistency. I try to use proper names in the code where possible... this is more a general remark rather then a request to change anything as you are consistent with the rest of the namings here

**Comment by jdsika** [comment]
I invited you to the org in order to create branches with "write rights"

File: src/config/yaml_baking_conf_parser.py [hunk 4]
```diff
@@ -176,6 +178,18 @@ def validate_min_delegation_amt(self, conf_obj):
                 )
             )
 
+    def validate_min_payment_amt(self, conf_obj):
+        if MIN_PAYMENT_AMT not in conf_obj:
+            conf_obj[MIN_PAYMENT_AMT] = 0
+            return
+
+        if not self.validate_non_negative_int(conf_obj[MIN_PAYMENT_AMT]):
```
Comment (jdsika) [inline src/config/yaml_baking_conf_parser.py:186]: It is the first time I become aware of the "validate_non_negative_int" function. That thing feels wrong... but not your cup of tea

File: src/configure.py [hunk 5]
```diff
@@ -195,9 +197,22 @@ def onmindelegation(input):
             input = "0"
         global parser
         parser.set(MIN_DELEGATION_AMT, float(input))
-        parser.validate_service_fee(parser.get_conf_obj())
```
Comment (jdsika) [inline src/configure.py:198]: Why are you deleting the check for "validate_service_fee"?

Comment (jdsika) [reply src/configure.py:198]: Before it felt like a "copy paste" error but now the previous check was removed intentionally?

File: src/configure.py [hunk 6]
```diff
@@ -419,7 +435,8 @@ def onprefinal(input):
             {"name": "go", "src": "foundersmap", "dst": "ownersmap"},
             {"name": "go", "src": "ownersmap", "dst": "mindelegation"},
             {"name": "go", "src": "mindelegation", "dst": "mindelegationtarget"},
-            {"name": "go", "src": "mindelegationtarget", "dst": "exclude"},
+            {"name": "go", "src": "mindelegationtarget", "dst": "minpayment"},
```
Comment (jdsika) [inline src/configure.py:438]: Am I seeing the "inheritance" here you mentioned in the docs?

File: docs/configuration.rst (hunk 3 above)
Comment (Vlad1mir-D) [reply docs/configuration.rst:130]: > I am having trouble understanding the "inherits behavior [...]" part. What exactly is inherited here? I am setting a value elsewhere and it is copied to min_payment_amt?

"inherits behavior" means there are no separate keyword like `mindelegation` available for "min payment threshold" in the `rules_map` allowing to set `TOF`/`TOB`/`TOE` specifically for `min_payment_amt` case so if payment excluded because of minimum payment threshold it will follow the rules set for `mindelegation`.

I foresee two options here:
* Put extended clarification into the documentation so it will be more or less obvious that `min_payment_amt` simply manipulates `min_delegation_amt` internally, follows all rules set for `mindelegation` in the `rules_map` and calculation won't reflect separate cause for payment excluded due to the `min_payment_amt` threshold.
* Implement separate `minpayment` key for the `rules_map` and add a separate exclusion reason which will properly reflect exclusion reason in the calculation report. Initially while taking another approach for `min_payment_amt` I did [implement all that stuff](https://github.com/Vlad1mir-D/tezos-reward-distributor/blob/minpayment-phased/src/config/yaml_baking_conf_parser.py#L404-L411) but then realized it could add unnecessary complexity to the configuration and decided to drop it in favor of making easier to review final implementation MR.

What do you think? Which is the better way to proceed?

File: src/configure.py (hunk 5 above)
Comment (Vlad1mir-D) [reply src/configure.py:198]: Yes, it's intentional as [`onmindelegation`](https://github.com/tezos-reward-distributor-organization/tezos-reward-distributor/blob/2c15b08aa15b92769176643f0b27e8ea507655eb/src/configure.py#L194-L204) itself seems to be a copy-paste from [`onservicefee`](https://github.com/tezos-reward-distributor-organization/tezos-reward-distributor/blob/2c15b08aa15b92769176643f0b27e8ea507655eb/src/configure.py#L139-L150).
Before this change [`onmindelegation`](https://github.com/tezos-reward-distributor-organization/tezos-reward-distributor/blob/2c15b08aa15b92769176643f0b27e8ea507655eb/src/configure.py#L194-L204) was validated by [`validate_service_fee`](https://github.com/tezos-reward-distributor-organization/tezos-reward-distributor/blob/2c15b08aa15b92769176643f0b27e8ea507655eb/src/config/yaml_baking_conf_parser.py#L163-L167) which is completely unrelated to `min_delegation_amt` in config but actually validates `service_fee` while proper validation [`validate_min_delegation_amt`](https://github.com/tezos-reward-distributor-organization/tezos-reward-distributor/blob/2c15b08aa15b92769176643f0b27e8ea507655eb/src/config/yaml_baking_conf_parser.py#L169-L179) for [`onmindelegation`](https://github.com/tezos-reward-distributor-organization/tezos-reward-distributor/blob/2c15b08aa15b92769176643f0b27e8ea507655eb/src/configure.py#L194-L204) exists but not used at all.

I fixed this issue but if you like I could create a separate MR for this fix but I prefer not to as the size of this change is relatively small.

File: src/config/yaml_baking_conf_parser.py (hunk 4 above)
Comment (Vlad1mir-D) [reply src/config/yaml_baking_conf_parser.py:186]: Ok, I'll mark this as resolved in this case.

File: src/configure.py (hunk 6 above)
Comment (Vlad1mir-D) [reply src/configure.py:438]: No, this part of code simply (well, not so _simple_ as it _should_ be in my opinion) sets the order of options during the configuration stage.
I explained what I meant by "inheritance" [here](https://github.com/tezos-reward-distributor-organization/tezos-reward-distributor/pull/613#discussion_r974822521).

**Comment by Vlad1mir-D** [comment]
> I invited you to the org in order to create branches with "write rights"

Thanks, I'll follow a basic procedure for all stuff I currently have in my TRD wrapper and try to move them into TRD one by one.

**Comment by Vlad1mir-D** [comment]
> Btw I think the abbreviations like "amt" for "amount" are a curse in the code. In the config we are basically forced to use it due to consistency. I try to use proper names in the code where possible... this is more a general remark rather then a request to change anything as you are consistent with the rest of the namings here

Noted. I attempted to use the same abbreviations everywhere but will use full length variable names in further features.

File: src/configure.py (hunk 5 above)
Comment (jdsika) [reply src/configure.py:198]: No. perfectly fine. You just see that I am basically reading these parts of the code for the first time :D

File: docs/configuration.rst (hunk 3 above)
Comment (jdsika) [reply docs/configuration.rst:130]: hm, I am not really a fan of implicit behavior - but the complexity is already high. I think I would opt for the minpayment key anyway and extend the docs for configuration for the user. It is also the thing that made me hesitate it he recursive loop that the min delegation is manipulated!
In addition the payment report is already incomplete (see #576 ) and I would like that fixed. Another puzzle with a double use of a keyword with implicit behaviour might not help there? What do you think?

Comment (jdsika) [reply docs/configuration.rst:130]: I understand correctly that otherwise the avoided payments would be marked with "min_delegation_amt" in the report? Do you think it is a false statement? Are there situations in which a min_payment_amt and a min_delegation_amt are not connected?
Thinking about it we could also couple the explanations of both configurations, write them under each other and explain the connection. In the morning I am leaning towards this :)

Comment (Vlad1mir-D) [reply docs/configuration.rst:130]: @jdsika 
> It is also the thing that made me hesitate it he recursive loop that the min delegation is manipulated!

I can attempt to fix https://github.com/Vlad1mir-D/tezos-reward-distributor/tree/minpayment-phased and implement this part without recursion but honestly I don't think there is any reason to feel any hesitation about recursion.

> In addition the payment report is already incomplete (see https://github.com/tezos-reward-distributor-organization/tezos-reward-distributor/issues/576 ) and I would like that fixed.

I'm not using dry-run mode because currently it requires an access to the signer and this destroys the whole idea of "dry run" as I'm afraid to provide an access to signer. So I can't comment this issue.

> Another puzzle with a double use of a keyword with implicit behaviour might not help there? What do you think?

I'm not sure what you mean by "help" so I think it definitely won't help with this thing :)

> I understand correctly that otherwise the avoided payments would be marked with "min_delegation_amt" in the report?

In current implementation - yes, they will be marked with `min_delegation_amt`.

> Do you think it is a false statement?

I would say this could be a confusing statement but there many other parts in TRD that's really confusing to me and this is among others is the reason I decided not to add an additional keyword.

> Are there situations in which a min_payment_amt and a min_delegation_amt are not connected?

What do you mean by "connected"? There could be situations where both `min_payment_amt` and `min_delegation_amt` will be defined in config but only one of them in current implementation could "win" (i.e. exclude most of the payments as the exclusion starts from the bottom) so I don't see any issues except confusion such report could represent but it could be addressed in the documentation.

> Thinking about it we could also couple the explanations of both configurations, write them under each other and explain the connection. In the morning I am leaning towards this :)

Well, it's possible to keep both reasons in the report, i.e. when payment is excluded due to one of the `minpayment` *or* `mindelegation` - there will be only one notice and when payment excluded by both thresholds there will be two notices.


To summarize:
1. Add an additional keyword `minpayment` to the `rules_map` with the same options which currently available for `mindelegation`
2. Keep both `mindelegation` and `minpayment` exclusion reasons in the report whenever payment get excluded

Are these two things going to be enough for this MR to be included into the master branch?
Thanks!

File: src/calc/phased_payment_calculator.py [hunk 7]
```diff
@@ -127,6 +132,32 @@ def calculate(self, reward_provider_model, adjustments=None):
             rwrd_logs, total_rwrd_amnt, adjustments
         )
 
+        # run phases again if calculated minimal payment amount is greater than configured
+        if self.min_payment_amnt and self.min_payment_amnt > int(
+            min([rl.adjusted_amount for rl in rwrd_logs if not rl.skipped])
```
Comment (vkresch) [inline src/calc/phased_payment_calculator.py:137]: Why do we need to run the calculation again if the configured min payment is greater then the calculated one? I thought we then should not pay at all and just return right? Isn't that the intention of this feature? Please correct me if I'm misunderstanding the feature.

**Comment by nicolasochem** [comment]
> I'm not using dry-run mode because currently it requires an access to the signer and this destroys the whole idea of "dry run" as I'm afraid to provide an access to signer. So I can't comment this issue.

Hi! A dry-run payout is a payout that has been calculated, signed, but not broadcasted. We exercise as much of the stack as possible (because it's useful) except we never inject the operation. There is no reason to be afraid (unless you are touching dry-run logic itself)

**Comment by jdsika** [comment]
> > I'm not using dry-run mode because currently it requires an access to the signer and this destroys the whole idea of "dry run" as I'm afraid to provide an access to signer. So I can't comment this issue.
> 
> Hi! A dry-run payout is a payout that has been calculated, signed, but not broadcasted. We exercise as much of the stack as possible (because it's useful) except we never inject the operation. There is no reason to be afraid (unless you are touching dry-run logic itself)

#615 should have just removed that feature

File: src/calc/phased_payment_calculator.py (hunk 7 above)
Comment (Vlad1mir-D) [reply src/calc/phased_payment_calculator.py:137]: > Isn't that the intention of this feature? Please correct me if I'm misunderstanding the feature.

Intention of this feature is to exclude all payments the same way as `min_delegation_amt` so this code works as expected.

File: docs/configuration.rst (hunk 3 above)
Comment (Vlad1mir-D) [reply docs/configuration.rst:130]: @jdsika What do you think?

**Comment by Vlad1mir-D** [comment]
@jdsika @vkresch Hi! What's left for this MR?

**Comment by Vlad1mir-D** [comment]
@vkresch It would be a pleasure to get a final list of notices that prevents this MR from merging until addressed as I'm really eager to donate all of changes I made in my own fork to the TRD community.

**Comment by Vlad1mir-D** [comment]
@vkresch 

> It would be easier to review the code if we could see a regression test e.g. test_min_payment_amount.py in the regression folder which tests your feature.

Because there is no response for https://github.com/tezos-reward-distributor-organization/tezos-reward-distributor/pull/613#discussion_r977091140 it would be better to get all notices at once as current implementation may change so do the tests and I prefer not to re-implement something multiple times if it's possible not to do so.

**Comment by Vlad1mir-D** [comment]
`/unstuck` :smiley_cat:

File: src/calc/phased_payment_calculator.py [hunk 8]
```diff
@@ -127,6 +132,32 @@ def calculate(self, reward_provider_model, adjustments=None):
...
+            logger.info(
+                "Setting min_delegation_amt to {:<,d} mutez due to min_payment_amt set to {:<,d}. Running calculations again.".format(
+                    self.min_delegation_amnt, self.min_payment_amnt
+                )
+            )
+            rwrd_logs, total_rwrd_amnt = self.calculate(
+                reward_provider_model, adjustments, True
+            )
+        elif rerun:
+            return rwrd_logs, total_rwrd_amnt
```
Comment (vkresch) [inline src/calc/phased_payment_calculator.py:159]: Please remove the return statement and the rerun variable since it is redundant. The function calculate returns anyway with reward_logs and the total reward amount. Ofc it sorts and logs in between but it should be fine.

File: src/calc/phased_payment_calculator.py [hunk 9]
```diff
@@ -51,15 +53,18 @@ def __init__(
     # owners reward = owners payment = total reward - delegators reward
     # founders reward = delegators fee = total reward - delegators reward
     ####
-    def calculate(self, reward_provider_model, adjustments=None):
+    def calculate(self, reward_provider_model, adjustments=None, rerun=False):
```
Comment (vkresch) [inline src/calc/phased_payment_calculator.py:56]: rerun is not needed

**Review by vkresch — CHANGES_REQUESTED** [review]
Thanks for the feature and sorry for the long review delay from my side. The implementation is good and the idea with the recalculation is a good example of code reusage :). The min payment is by default 0 if not defined in a config thus the feature is backwards compatible. I have just some minor change request and after that it can be merged. Thanks again and sorry for the inconvenience.

**Review by jdsika — APPROVED** [review]
